### ISSUE TITLE
Getting a stack trace when running example args

### ISSUE THREAD

**Issue by danielhoherd** [robsonbittencourt/aws-cost-miner#1]
When running the example from README I am encountering a stack trace. I only changed the way the PWD is referenced and the groupBy field. My csv is 45 MB and 51888 lines.

```
$ docker run --rm --name aws-cost-miner -v ${PWD}/CostReport-1.csv:/app/data.csv -e report=EC2_COST_REPORT -e groupBy="resourceTags/user:Project" robsonbittencourt/aws-cost-miner
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'awsCostMinerApplication': Invocation of init method failed; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:139)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:414)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:845)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:744)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:391)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1204)
	at com.rbittencourt.aws.cost.miner.AwsCostMinerApplication.main(AwsCostMinerApplication.java:22)
	... 8 more
Caused by: java.lang.NullPointerException
	at com.rbittencourt.aws.cost.miner.infrastructure.file.ReservedInstanceInfoCsvRepository.lambda$findReservedInstanceInfos$0(ReservedInstanceInfoCsvRepository.java:27)
	at com.rbittencourt.aws.cost.miner.infrastructure.file.CsvReader.csvToObjects(CsvReader.java:42)
	at com.rbittencourt.aws.cost.miner.infrastructure.file.ReservedInstanceInfoCsvRepository.findReservedInstanceInfos(ReservedInstanceInfoCsvRepository.java:28)
	at com.rbittencourt.aws.cost.miner.infrastructure.file.BillingInfoCsvRepository.findBillingInfos(BillingInfoCsvRepository.java:24)
	at com.rbittencourt.aws.cost.miner.domain.miner.DataOrganizer.organizeData(DataOrganizer.java:30)
	at com.rbittencourt.aws.cost.miner.domain.miner.AwsCostMinerImpl.miningCostData(AwsCostMinerImpl.java:26)
	at com.rbittencourt.aws.cost.miner.application.report.ConsoleCostReport.writeReport(ConsoleCostReport.java:31)
	at com.rbittencourt.aws.cost.miner.AwsCostMinerApplication.start(AwsCostMinerApplication.java:18)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:363)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:307)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:136)
	... 25 more
```

```
$ docker-images robsonbittencourt/aws-cost-miner
robsonbittencourt/aws-cost-miner:latest
    .ID:            6532ff9d12da
    .Repository:    robsonbittencourt/aws-cost-miner
    .Tag:           latest
    .Digest:        sha256:8cc2cfd3f6d5fefc53d6ca01aa802fe0b3aa3b7cc42d321fbc888c4b266f3923
    .CreatedSince:  44 hours ago
    .CreatedAt:     2020-02-04 17:08:53 -0800 PST
    .Size:          485MB
```

**Comment by robsonbittencourt** [comment]
Hi @danielhoherd, thanks for your feedback.

Unfortunately I built this tool, having only one CSV billing file available. Because of this, some surprises like these may appear.

At first the error seems to be due to some information in the CSV that I did not foresee. I made a new version trying to deal with your case. Please update your image to the latest version, or use the 1.0.1 tag, and test again.

If possible, I ask you two favors:

- Check if your file has any lines with a blank UsageType column.

- Enter an example of your file with a few lines. Feel free to omit sensitive information, the most important is the structure.

**Comment by danielhoherd** [comment]
Slightly different error with the latest build.

```
$ docker-images robsonbittencourt/aws-cost-miner
robsonbittencourt/aws-cost-miner:latest
    .ID:            d6c8e7cdfbf1
    .Repository:    robsonbittencourt/aws-cost-miner
    .Tag:           latest
    .Digest:        sha256:8d50ce0e11cf6693fb3400d782c16d822266d19a443ca9fe64ab6bd22a3c7996
    .CreatedSince:  18 hours ago
    .CreatedAt:     2020-02-06 15:13:28 -0800 PST
    .Size:          485MB
```

```
$ docker run --rm --name aws-cost-miner \
  -v ${PWD}/CostReport-1.csv:/app/data.csv \
  -e report=EC2_COST_REPORT \
  -e groupBy="resourceTags/user:Project" \
  robsonbittencourt/aws-cost-miner
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'awsCostMinerApplication': Invocation of init method failed; nested exception is java.lang.OutOfMemoryError: Java heap space
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:139)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:414)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:845)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:744)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:391)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1204)
	at com.rbittencourt.aws.cost.miner.AwsCostMinerApplication.main(AwsCostMinerApplication.java:22)
	... 8 more
Caused by: java.lang.OutOfMemoryError: Java heap space
	at jdk.internal.reflect.GeneratedMethodAccessor7.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at com.fasterxml.jackson.databind.introspect.AnnotatedMethod.callOnWith(AnnotatedMethod.java:122)
	at com.fasterxml.jackson.databind.deser.SettableAnyProperty.set(SettableAnyProperty.java:176)
	at com.fasterxml.jackson.databind.deser.SettableAnyProperty.deserializeAndSet(SettableAnyProperty.java:134)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1561)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:294)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.MappingIterator.nextValue(MappingIterator.java:277)
	at com.fasterxml.jackson.databind.MappingIterator.next(MappingIterator.java:192)
	at com.rbittencourt.aws.cost.miner.infrastructure.file.CsvReader.csvToObjects(CsvReader.java:40)
	at com.rbittencourt.aws.cost.miner.infrastructure.file.CsvReader.csvToObjects(CsvReader.java:28)
	at com.rbittencourt.aws.cost.miner.infrastructure.file.BillingInfoCsvRepository.findBillingInfos(BillingInfoCsvRepository.java:26)
	at com.rbittencourt.aws.cost.miner.domain.miner.DataOrganizer.organizeData(DataOrganizer.java:30)
	at com.rbittencourt.aws.cost.miner.domain.miner.AwsCostMinerImpl.miningCostData(AwsCostMinerImpl.java:26)
	at com.rbittencourt.aws.cost.miner.application.report.ConsoleCostReport.writeReport(ConsoleCostReport.java:31)
	at com.rbittencourt.aws.cost.miner.AwsCostMinerApplication.start(AwsCostMinerApplication.java:18)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:363)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:307)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:136)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:414)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.AbstractBeanFactory$$Lambda$160/0x0000000801233c40.getObject(Unknown Source)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
```

I actually do not have a "UsageType" column, but I have "product/usagetype" and "lineItem/UsageType". I have attached the first two lines of the csv, with some sensitive information replaced with repeating strings.
[head-n-2-CostReport-1.csv.gz](https://github.com/robsonbittencourt/aws-cost-miner/files/4172120/head-n-2-CostReport-1.csv.gz)

**Comment by robsonbittencourt** [comment]
@danielhoherd with your sample file I was able to understand the problem.

I built this tool based on reports that I had. And these kind of reports have been depreciated by AWS.

As can be seen in the links below, there is a new type of report that has been adopted by AWS. And this new report is your model.

I configured the creation of a report of this new model in my account, and in the next few days I will try to adapt aws-cost-miner for this model.

https://docs.aws.amazon.com/cur/latest/userguide/detailed-billing.html

https://docs.aws.amazon.com/cur/latest/userguide/detailed-billing-migrate.html

**Comment by robsonbittencourt** [comment]
@danielhoherd I made the necessary adaptations so that aws-cost-miner works with both aws reports mentioned above, which includes yours.

Update your Docker image to the latest version or use the 1.1.0 tag and do a new test.

Please keep me informed if it worked.

**Comment by danielhoherd** [comment]
I unzipped a fresh csv file and ran 1.1.0 on my laptop, and noticed that the error I got this time and the previous one in this thread is OOM. I copied the 200M csv to a machine with 251GB of memory and ran 1.1.0 on it and got a different error.

```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:           251G         42G        4.6G        8.2G        204G        198G
Swap:            0B          0B          0B
$ docker run --rm --name aws-cost-miner -v ${PWD}/CostReport-1.csv:/app/data.csv -e report=EC2_COST_REPORT -e groupBy="resourceTags/user:Project" robsonbittencourt/aws-cost-miner:1.1.0
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'awsCostMinerApplication': Invocation of init method failed; nested exception is java.lang.StringIndexOutOfBoundsException
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:139)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:414)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:593)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:515)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:320)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:318)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:845)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:877)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:744)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:391)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1204)
	at com.rbittencourt.aws.cost.miner.AwsCostMinerApplication.main(AwsCostMinerApplication.java:26)
	... 8 more
Caused by: java.lang.StringIndexOutOfBoundsException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:919)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.rbittencourt.aws.cost.miner.domain.billing.BillingInfos.groupBy(BillingInfos.java:52)
	at com.rbittencourt.aws.cost.miner.domain.metric.CostByGroup.metricValues(CostByGroup.java:31)
	at com.rbittencourt.aws.cost.miner.domain.metric.CostByInstanceFamilyMetric.calculateMetric(CostByInstanceFamilyMetric.java:25)
	at com.rbittencourt.aws.cost.miner.domain.miner.AwsCostMinerImpl.getMetricResults(AwsCostMinerImpl.java:46)
	at com.rbittencourt.aws.cost.miner.domain.miner.AwsCostMinerImpl.minedData(AwsCostMinerImpl.java:35)
	at com.rbittencourt.aws.cost.miner.domain.miner.AwsCostMinerImpl.miningCostData(AwsCostMinerImpl.java:28)
	at com.rbittencourt.aws.cost.miner.application.report.ConsoleCostReport.writeReport(ConsoleCostReport.java:31)
	at com.rbittencourt.aws.cost.miner.AwsCostMinerApplication.start(AwsCostMinerApplication.java:19)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:363)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:307)
	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:136)
	... 25 more
Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end -1, length 13
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3410)
	at java.base/java.lang.String.substring(String.java:1883)
	at com.rbittencourt.aws.cost.miner.domain.billing.BillingInfo.instanceFamily(BillingInfo.java:199)
	at java.base/java.util.stream.Collectors.lambda$groupingBy$53(Collectors.java:1127)
	at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:952)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:926)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```

**Comment by robsonbittencourt** [comment]
@danielhoherd we've made progress. Analyzing the stacktrace this time at least read your file. So the structure problem is ok. Now this error that you sent has to do with the data. It is a little difficult for me to create a logic to extract certain data, without knowing all the possibilities that AWS can generate. The error occurred in an attempt to use substring to extract the family name from the VM. I did the treatment so that if it is not possible to extract, use a default value informing that the family is unknown.

I ask you to update your image (latest or 1.1.1) and test again. I believe we are close to the solution.

About memory problems I use aws-cost-miner on two machines. One with 12gb of RAM and another with 16gb and I never had problems with OutOfMemory. The program actually consumes relatively much memory (~ 3GB for a 300MB CSV) but for a short time. But after we solve this functional problem, I can review the strategy of reading the file, to optimize the use of memory.
